src/aspire/source/micrograph.py

```diff
@@ -0,0 +1,306 @@
+        """
+        String representation of the MicrographSource.
+        """
+        return f"{self.micrograph_count} {self.dtype.name} micrographs of size {self.micrograph_size}x{self.micrograph_size}"
+
+    def __len__(self):
+        """
+        Returns the number of micrographs.
+        """
+        return self.micrograph_count
+
+    @property
+    def clean_micrographs(self):
+        """
+        Returns the micrographs without any noise.
+        """
+        return self._clean_micrographs_accessor
+
+    @property
+    def micrographs(self):
+        """
+        Returns the micrographs with any added noise.
+        """
+        return self._micrographs_accessor
+
+    def _micrographs(self, indices):
+        """
+        Accesses and returns micrographs with any added noise.
+        :param indices: A 1-D NumPy array of integer indices.
+        :return: An `Image` object representing the clean micrograph.
+        """
+        micrographs = self._clean_micrographs(indices)
+        if self.noise_adder:
+            micrographs = self.noise_adder.forward(micrographs)
+        return micrographs
+
+    def _clean_micrographs(self, indices):
+        """
+        Accesses and returns micrographs without any added noise.
+
+        :param indices: A 1-D NumPy array of integer indices.
+        :return: An `Image` object representing the clean micrograph
+        """
+        # Initialize empty micrograph
+        n_micrographs = len(indices)
+        clean_micrograph = np.zeros(
+            (n_micrographs, self.micrograph_size, self.micrograph_size),
+            dtype=self.dtype,
+        )
+        # Pad the micrograph
+        clean_micrograph = np.pad(
+            clean_micrograph,
+            ((0, 0), (self.pad, self.pad), (self.pad, self.pad)),
+            "constant",
+            constant_values=(0),
+        )
+        # Get centers
+        parity = self.particle_box_size % 2
+        for m in range(n_micrographs):
+            global_id = indices[m]
+            image = self.simulation.clean_images[self.get_particle(global_id)]
+            centers = self.centers[global_id]
+            x_lefts = centers[:, 0] - self.particle_box_size // 2 + self.pad
+            x_rights = centers[:, 0] + self.particle_box_size // 2 + parity + self.pad
+            y_lefts = centers[:, 1] - self.particle_box_size // 2 + self.pad
+            y_rights = centers[:, 1] + self.particle_box_size // 2 + parity + self.pad
+            # Subtract image from the micrograph using the particle's bounding box and center
+            for p in range(self.particles_per_micrograph):
+                clean_micrograph[m][
+                    x_lefts[p] : x_rights[p], y_lefts[p] : y_rights[p]
+                ] = (
+                    clean_micrograph[m][
+                        x_lefts[p] : x_rights[p], y_lefts[p] : y_rights[p]
+                    ]
+                    - image[p]
+                )
+            # Normalize micrographs
```

Comment:
I'm not sure we want to do this. Again, normalization may be handled in the `Simulation` it self.